indigo_app/js/components/DocumentTOCView.vue

```diff
@@ -0,0 +1,297 @@
+        console.log('rebuilding TOC');
+        const oldLength = this.toc.length,
+            index = this.backBoneContext.selection.get('index');
+
+        this.buildToc();
+
+        if (index > this.toc.length-1) {
+          // we've selected past the end of the TOC
+          this.selectItem(this.toc.length-1);
+
+        } else if (force || (index > -1 && this.toc.length !== oldLength)) {
+          // arrangament of the TOC has changed, re-select the item we want
+          this.selectItem(index, true);
+
+        } else {
+          if (index > -1) {
+            this.toc[index].selected = true;
+          }
+
+          this.render();
+        }
+      }
+    },
+
+    buildToc () {
+      // Get the table of contents of this document
+      // roots is a list of the root elements of the toc tree
+      // toc is an ordered list of all items in the toc
+      const roots = [],
+          toc = [];
+      const tradition = Indigo.traditions.get(this.backBoneContext.model.document.get('country'));
+
+      const iterateChildren = (node, parentItem) => {
+        const kids = node.children;
+
+        for (let i = 0; i < kids.length; i++) {
+          let kid = kids[i],
+              tocItem = null;
+
+          if (tradition.is_toc_deadend(kid)) continue;
+
+          if (tradition.is_toc_element(kid)) {
+            tocItem = generateToc(kid);
+            tocItem.index = toc.length;
+            toc.push(tocItem);
+
+            if (parentItem) {
+              parentItem.children = parentItem.children || [];
+              parentItem.children.push(tocItem);
+            } else {
+              roots.push(tocItem);
+            }
+          }
+
+          iterateChildren(kid, tocItem || parentItem);
+        }
+      }
+
+      const getHeadingText = (node) => {
+        let headingText = '';
+        const result = this.backBoneContext.model.xpath('./a:heading//text()[not(ancestor::a:authorialNote)]', node);
+
+        for (let i = 0; i < result.snapshotLength; i++) {
+          headingText += result.snapshotItem(i).textContent;
+        }
+
+        return headingText;
+      }
+
+      function generateToc (node) {
+        const $node = $(node);
+        const $component = $node.closest('attachment');
+        let qualified_id = node.getAttribute('eId');
+
+        if ($component.length > 0) {
+          qualified_id = $component.attr('eId') + '/' + qualified_id;
+        }
+
+        const item = {
+          'num': $node.children('num').text(),
+          'heading': getHeadingText(node),
+          'element': node,
+          'type': node.localName,
+          'id': qualified_id,
+        };
+        item.title = tradition.toc_element_title(item);
+        return item;
+      }
+
+      iterateChildren(this.backBoneContext.model.xmlDocument);
+
+      this.toc = toc;
+      this.roots = roots;
+
+      this.mergeIssues();
+    },
+
+    issuesChanged () {
+      this.mergeIssues();
+      this.render();
```

Comment:
in theory, Vue should detect that `mergeIssues` has updated the issues linked to each item in the toc, and re-render. So calling `render()` falls away.

Reply:
refactored